### PR TITLE
Wpd 146

### PR DESCRIPTION
**Pull request by d-wysocki**
image

image

https://kindergarten.stxnext.pl/browse/WDP-146

### CODE REVIEW

File: src/presence_analyzer/static/js/overtime.js [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    return result;
+}
+
+(function($) {
+    $(document).ready(function(){
+        var $loading = $('#loading'),
+            $chartDiv = $("#chart_div");
+
+            $loading.hide();
+            $chartDiv.show();
```
Comment (piotrgrundas) [inline src/presence_analyzer/static/js/overtime.js]: indent

File: src/presence_analyzer/static/js/overtime.js [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+        $.getJSON('/api/v1/overtime/', function(result) {
+            var finalResult = [],
+                chart = new google.visualization.Timeline($chartDiv[0]),
+                dataTable = new google.visualization.DataTable(),
+                options = {hAxis: { format: 'd' }};
+
+            $.each(result, function(day, intervals) {
+                finalResult.push([
+                    (day+1).toString(),
```
Comment (piotrgrundas) [inline src/presence_analyzer/static/js/overtime.js]: space around `+`

File: src/presence_analyzer/static/js/overtime.js [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+                    intervals[1]['name'].toString(),
+                    new Date(0,0,0,0,0,0),
```
Comment (piotrgrundas) [inline src/presence_analyzer/static/js/overtime.js]: ditto

File: src/presence_analyzer/utils.py [hunk 4]
```diff
@@ -199,3 +199,41 @@ def star_end_time(data, user_id):
...
+                    result[user] = {
+                        'name': names[str(user)]['name'],
+                        'overtime': []
+                    }
+
+                result[user]['overtime'].append(overtime - 28800)
+        try:
+            result[user]['overtime'] = sum(result[user]['overtime'])
+        except KeyError:
+            pass
```
Comment (piotrgrundas) [inline src/presence_analyzer/utils.py]: enter

File: src/presence_analyzer/templates/overtime.html [hunk 5]
```diff
@@ -0,0 +1,5 @@
+<%inherit file="base.html"/>
+
+    <%block name='head'>
+        <script type="text/javascript" src="${ url_for('static', filename='js/overtime.js') }"></script>
+    </%block>
```
Comment (piotrgrundas) [inline src/presence_analyzer/templates/overtime.html]: ???

File: src/presence_analyzer/utils.py [hunk 6]
```diff
@@ -199,3 +199,41 @@ def star_end_time(data, user_id):
...
+        for item in data[user]:
+            start = seconds_since_midnight(
+                data[user][item]['start']
+            )
+            end = seconds_since_midnight(
+                data[user][item]['end']
+            )
+            overtime = end - start
+
+            if overtime > 28800:
```
Comment (piotrgrundas) [inline src/presence_analyzer/utils.py]: `28800` store this in a descriptive var

File: src/presence_analyzer/utils.py [hunk 7]
```diff
@@ -199,3 +199,41 @@ def star_end_time(data, user_id):
...
+
+def get_overtime(data):
+    """
+    Calculate user overtime and sort it.
+    """
+    result = {}
+    names = get_user()
+
+    for user in data:
+        for item in data[user]:
```
Comment (piotrgrundas) [inline src/presence_analyzer/utils.py]: 1 loop!

**Review by piotrgrundas — CHANGES_REQUESTED** [review]

File: src/presence_analyzer/utils.py (hunk 6 above)
Comment (dpolcyn) [reply src/presence_analyzer/utils.py]: Why 28800?

File: src/presence_analyzer/static/js/overtime.js (hunk 2 above)
Comment (dpolcyn) [reply src/presence_analyzer/static/js/overtime.js]: why you sum this here?

File: src/presence_analyzer/tests.py [hunk 8]
```diff
@@ -78,6 +78,23 @@ def test_avatar_api(self):
...
+        data = json.loads(resp.data)
+
+        self.assertEqual(resp.status_code, 200)
+        self.assertEqual(resp.content_type, 'application/json')
+        self.assertEqual(
+            data, [
+                [176, {'overtime': [1801], 'name': 'Adrian K.'}],
+                [141, {'overtime': 60, 'name': 'Adam P.'}]
+            ]
+
```
Comment (dpolcyn) [inline src/presence_analyzer/tests.py]: delete

File: src/presence_analyzer/views.py [hunk 9]
```diff
@@ -157,3 +159,10 @@ def presence_start_end(user_id):
         log.debug('User %s not found!', user_id)
         abort(404)
     return star_end_time(data, user_id)
+
+
+@app.route('/api/v1/overtime/', methods=['GET'])
+@jsonify
+def overtime():
+    data = get_data()
+    return get_overtime(data)
```
Comment (dpolcyn) [inline src/presence_analyzer/views.py]: return get_overtime(get_data()) or get_data move to get_overtime

File: src/presence_analyzer/static/js/overtime.js [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+    return result;
+}
+
+(function($) {
+    $(document).ready(function(){
+        var $loading = $('#loading'),
+            $chartDiv = $("#chart_div");
+
+                $loading.hide();
+                $chartDiv.show();
```
Comment (piotrgrundas) [inline src/presence_analyzer/static/js/overtime.js]: ?

File: src/presence_analyzer/templates/mean_time_weekday.html [hunk 11]
```diff
@@ -1,8 +1,8 @@
...
+<%inherit file="base.html" />
 
-    <%block name='head'>
+    <%block name='head' >
         <script type="text/javascript">
             google.load("visualization", "1", {packages:["corechart"], 'language': 'pl'});
         </script>
         <script type="text/javascript" src="${ url_for('static', filename='js/mean_time.js') }"></script>
-    </%block>
+    </% block >
```
Comment (piotrgrundas) [inline src/presence_analyzer/templates/mean_time_weekday.html]: ? Stick to one convention

**Review by piotrgrundas — CHANGES_REQUESTED** [review]

**Comment by piotrgrundas** [comment]
Correct previous comments

File: src/presence_analyzer/static/js/overtime.js [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+            });
+
+            dataTable.addColumn({ type: 'string', id: 'Position' });
+            dataTable.addColumn({ type: 'string', id: 'Name' });
+            dataTable.addColumn({ type: 'date', id: 'Start' });
+            dataTable.addColumn({ type: 'date', id: 'End' });
+            dataTable.addRows(finalResult);
+
+            chart.draw(dataTable, options);
+
```
Comment (dpolcyn) [inline src/presence_analyzer/static/js/overtime.js]: remove

File: src/presence_analyzer/static/js/overtime.js [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+        })
+            .fail(function() {
```
Comment (dpolcyn) [inline src/presence_analyzer/static/js/overtime.js]: check indent, not sure

File: src/presence_analyzer/templates/mean_time_weekday.html (hunk 11 above)
Comment (dpolcyn) [reply src/presence_analyzer/templates/mean_time_weekday.html]: :+1:

File: src/presence_analyzer/utils.py [hunk 14]
```diff
@@ -199,3 +199,65 @@ def star_end_time(data, user_id):
...
+    date = datetime(
+        year,
+        month,
+        calendar.monthrange(year, month)[1]
+    )
+
+    return sum([
+        1 for x in range(1, date.day)
+        if datetime(year, month, x).weekday() < 5
+    ]) * eight_hour
```
Comment (dpolcyn) [inline src/presence_analyzer/utils.py]: make no sense, multiple it by 28800 or move it to settings and call 'working_hours'

File: src/presence_analyzer/utils.py [hunk 15]
```diff
@@ -199,3 +199,65 @@ def star_end_time(data, user_id):
...
 
     return result
+
+
+def bussines_days(year_month):
+    """
+    Calculate soconds of worked days in month.
+    """
+    year, month = year_month
+    eight_hour = 28800
```
Comment (dpolcyn) [inline src/presence_analyzer/utils.py]: this isn't clear if u don't check if it is 8 hours, no comment etc... 
thats how you should decrare it:
xxx = 60*60*8

File: src/presence_analyzer/utils.py [hunk 16]
```diff
@@ -199,3 +199,65 @@ def star_end_time(data, user_id):
...
+        if datetime(year, month, x).weekday() < 5
+    ]) * eight_hour
+
+
+def get_overtime(data):
+    """
+    Calculate user overtime and sort it.
+    """
+    names = get_user()
+    result = {}
```
Comment (dpolcyn) [inline src/presence_analyzer/utils.py]: im not sure if you need 2 dicts here to make this, try using one

File: src/presence_analyzer/tests.py [hunk 17]
```diff
@@ -419,6 +437,34 @@ def test_star_end_time(self):
...
+                    'name': 'Adrian K.'
+                }
+            )]
+        )
+
+    def test_bussines_day(self):
+        """
+        Test for worked seconds in month.
+        """
+        self.assertEqual(
```
Comment (dpolcyn) [inline src/presence_analyzer/tests.py]: read again pep8 rules, I think u don't understand it yet

File: src/presence_analyzer/utils.py [hunk 18]
```diff
@@ -199,3 +199,70 @@ def star_end_time(data, user_id):
...
+                'name': names[str(user)]['name'],
+                'overtime': finally_overtime
+            }
+        try:
+            if 'name' not in result[user][year_month]:
+                del result[user]
+        except KeyError:
+            pass
+
+    return sorted(
```
Comment (dpolcyn) [inline src/presence_analyzer/utils.py]: Won't work properly

File: src/presence_analyzer/views.py [hunk 19]
```diff
@@ -149,11 +135,16 @@ def presence_weekday_view(user_id):
 @app.route('/api/v1/presence_start_end/<int:user_id>', methods=['GET'])
 @jsonify
 def presence_start_end(user_id):
-    """
-    Return average time when user start the work and when user end the work.
+    """Return average time when user start the work and when user end the work.
```
Comment (dpolcyn) [inline src/presence_analyzer/views.py]: Now is wrong

Comment (dpolcyn) [reply src/presence_analyzer/views.py]: Just go with 
"""
xxx
"""

File: src/presence_analyzer/templates/overtime.html [hunk 20]
```diff
@@ -0,0 +1,5 @@
+<%inherit file="base.html"/>
+
+    <%block name='head'>
+        <script src="${ url_for('static', filename='js/overtime.js') }"></script>
+    </%block>
```
Comment (piotrgrundas) [inline src/presence_analyzer/templates/overtime.html]: Really? Bocks shouldn't be indend! It's content should be!

File: src/presence_analyzer/utils.py [hunk 21]
```diff
@@ -199,3 +193,70 @@ def star_end_time(data, user_id):
...
+    """
+    year, month = year_month
+    date = datetime(
+        year,
+        month,
+        calendar.monthrange(year, month)[1]
+    )
+
+    return sum([
+        1 for x in range(1, date.day)
```
Comment (piotrgrundas) [inline src/presence_analyzer/utils.py]: `xrange`

File: src/presence_analyzer/utils.py [hunk 22]
```diff
@@ -199,3 +193,70 @@ def star_end_time(data, user_id):
...
+    """
+    Calculate user overtime and sort it.
+    """
+    names = get_user()
+    result = {}
+    for user in data:
+        if str(user) not in names.keys():
+            continue
+        for item in data[user]:
+
```
Comment (piotrgrundas) [inline src/presence_analyzer/utils.py]: remove

File: src/presence_analyzer/utils.py [hunk 23]
```diff
@@ -199,3 +193,70 @@ def star_end_time(data, user_id):
...
+                data[user][item]['start']
+            )
+            end = seconds_since_midnight(
+                data[user][item]['end']
+            )
+            overtime = end - start
+            year_month = (item.year, item.month)
+
+            if user not in result or year_month not in result[user]:
+                result.setdefault(user, {})[year_month] = {'overtime': []}
```
Comment (piotrgrundas) [inline src/presence_analyzer/utils.py]: Why not just
`result[user] = {year_month: {'overtime': []}}`

**Review by piotrgrundas — CHANGES_REQUESTED** [review]

File: src/presence_analyzer/utils.py [hunk 24]
```diff
@@ -166,20 +167,15 @@ def average_seconds(data, board):
     """
     Calculate average time in seconds and return it as string.
     """
-    return str(
-        timedelta(seconds=mean(data[board]))
-    ).split(".")[0]
+    return str(timedelta(seconds=mean(data[board]))).split(".")[0]
```
Comment (piotrgrundas) [inline src/presence_analyzer/utils.py]: `'`
IMO previous version was more readable

**Review by piotrgrundas — CHANGES_REQUESTED** [review]

**Pull request merged**